// @narative/gatsby-theme-novela/src/gatsby/node/createPages.js
/* eslint-disable no-console, import/no-extraneous-dependencies, prefer-const, no-shadow */

require('dotenv').config();

const log = (message, section) =>
  console.log(`\n\u001B[36m${message} \u001B[4m${section}\u001B[0m\u001B[0m\n`);

const path = require('path');
const createPaginatedPages = require('gatsby-paginate');

const templatesDirectory = path.resolve(__dirname, '../../templates');
const templates = {
  articles: path.resolve(templatesDirectory, 'articles.template.tsx'),
  article: path.resolve(templatesDirectory, 'article.template.tsx'),
  author: path.resolve(templatesDirectory, 'author.template.tsx'),
  tag: path.resolve(templatesDirectory, "tag.template.tsx"),
  ampPage: path.resolve(templatesDirectory, "article.template.amp.tsx"),
};

const query = require('../data/data.query');
const normalize = require('../data/data.normalize');

// ///////////////// Utility functions ///////////////////

function buildPaginatedPath(index, basePath) {
  if (basePath === '/') {
    return index > 1 ? `${basePath}page/${index}` : basePath;
  }
  return index > 1 ? `${basePath}/page/${index}` : basePath;
}

function slugify(string, base) {
  const slug = string
    .toLowerCase()
    .normalize('NFD')
    .replace(/[\u0300-\u036F]/g, '')
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/(^-|-$)+/g, '');

  return `${base}/${slug}`.replace(/\/\/+/g, '/');
}

function getUniqueListBy(array, key) {
  return [...new Map(array.map(item => [item[key], item])).values()];
}

const byDate = (a, b) => new Date(b.dateForSEO) - new Date(a.dateForSEO);

// ///////////////////////////////////////////////////////

module.exports = async ({ actions: { createPage }, graphql }, themeOptions) => {
  const {
    rootPath,
    basePath = '/',
    authorsPath = '/authors',
    authorsPage = true,
    tagsPath = '/tags',
    tagsPage = true,
    pageLength = 6,
    sources = {},
    mailchimp = '',
  } = themeOptions;

  const { data } = await graphql(`
    query siteQuery {
      site {
        siteMetadata {
          siteUrl
        }
      }
    }
  `);

  console.log(sources);
  // Defaulting to look at the local MDX files as sources.
  const { local = true, contentful = false } = sources;

  let authors;
  let articles;
  let tags;

  const dataSources = {
    local: { authors: [], articles: [], tags: [] },
    contentful: { authors: [], articles: [], tags: [] },
    netlify: { authors: [], articles: [], tags: [] },
  };

  if (rootPath) {
    log('Config rootPath', rootPath);
  } else {
    log('Config rootPath not set, using basePath instead =>', basePath);
  }

  log('Config basePath', basePath);
  if (authorsPage) log('Config authorsPath', authorsPath);
  if (tagsPage) log('Config tagsPath', tagsPath);

  if (local) {
    try {
      log('Querying Authors, Articles, Tags & Pages source:', 'Local');
      const localAuthors = await graphql(query.local.authors);
      const localArticles = await graphql(query.local.articles);
      const localTags = await graphql(query.local.tags);

      dataSources.local.authors = localAuthors.data.authors.edges.map(
        normalize.local.authors,
      );

      dataSources.local.articles = localArticles.data.articles.edges.map(
        normalize.local.articles,
      );

      // dataSources.local.tags = localTags.data.tags.edges.map(
      //   normalize.local.tags,
      // );

      dataSources.local.tags = localTags.data.tags.edges.map((tag) => tag.node);

    } catch (error) {
      console.error(error);
    }
  }

  if (contentful) {
    try {
      log('Querying Authors, Articles, Tags & Pages source:', 'Contentful');
      const contentfulAuthors = await graphql(query.contentful.authors);
      const contentfulArticles = await graphql(query.contentful.articles);
      const contentfulTags = await graphql(query.contentful.tags);

      dataSources.contentful.authors = contentfulAuthors.data.authors.edges.map(
        normalize.contentful.authors,
      );

      dataSources.contentful.articles = contentfulArticles.data.articles.edges.map(
        normalize.contentful.articles,
      );

      dataSources.contentful.tags = contentfulTags.data.tags.edges.map(
        normalize.contentful.tags,
      );

    } catch (error) {
      console.error(error);
    }
  }

  // Combining together all the articles from different sources
  articles = [
    ...dataSources.local.articles,
    ...dataSources.contentful.articles,
    ...dataSources.netlify.articles,
  ].sort(byDate);

  const articlesThatArentSecret = articles.filter(article => !article.secret);

  // Combining together all the authors from different sources
  authors = getUniqueListBy(
    [
      ...dataSources.local.authors,
      ...dataSources.contentful.authors,
      ...dataSources.netlify.authors,
    ],
    'name',
  );
  // Combining together all the tags from different sources
  tags = getUniqueListBy(
    [
      ...dataSources.local.tags,
      ...dataSources.contentful.tags,
      ...dataSources.netlify.tags,
    ],
    'name',
  );

  // if (articles.length === 0 || authors.length === 0 || tags.length === 0) {
  //   throw new Error(`
  //   You must have at least one Author, Post and Tag. As reference you can view the
  //   example repository. Look at the content folder in the example repo.
  //   https://github.com/narative/gatsby-theme-novela-example
  // `);
  // }

  if (articles.length === 0 || authors.length === 0) {
    throw new Error(`
    You must have at least one Author, Post and Tag. As reference you can view the
    example repository. Look at the content folder in the example repo.
    https://github.com/narative/gatsby-theme-novela-example
  `);
  }

  /**
   * Once we've queried all our data sources and normalized them to the same structure
   * we can begin creating our pages. First, we'll want to create all main articles pages
   * that have pagination.
   * /articles
   * /articles/page/1
   * ...
   */
  log('Creating', 'articles page');
  createPaginatedPages({
    edges: articlesThatArentSecret,
    pathPrefix: basePath,
    createPage,
    pageLength,
    pageTemplate: templates.articles,
    buildPath: buildPaginatedPath,
    context: {
      authors,
      tags,
      basePath,
      skip: pageLength,
      limit: pageLength,
    },
  });

  /**
   * Once the list of articles have bene created, we need to make individual article posts.
   * To do this, we need to find the corresponding authors and tags since we allow for co-authors and multiple tags.
   */
  log('Creating', 'article posts');
  articles.forEach((article, index) => {
    // Match the Author to the one specified in the article
    let authorsThatWroteTheArticle;
    let tagsThatWroteTheArticle;
    try {
      authorsThatWroteTheArticle = authors.filter(author => {
        const allAuthors = article.author
          .split(',')
          .map(a => a.trim().toLowerCase());

        return allAuthors.some(a => a === author.name.toLowerCase());
      });
    } catch (error) {
      throw new Error(`
        We could not find the Author for: "${article.title}".
        Double check the author field is specified in your post and the name
        matches a specified author.
        Provided author: ${article.author}
        ${error}
      `);
    }
    // Match the Tag to the one specified in the article
    try {
      tagsThatWroteTheArticle = tags.filter(tag => {
        const allTags = article.tag
          .split(',')
          .map(a => a.trim().toLowerCase());

        return allTags.some(a => a === tag.name.toLowerCase());
      });
    } catch (error) {
      throw new Error(`
        We could not find the Tag for: "${article.title}".
        Double check the tag field is specified in your post and the name
        matches a specified tag.
        Provided tag: ${article.tag}
        ${error}
      `);
    }

    /**
     * We need a way to find the next articles to suggest at the bottom of the articles page.
     * To accomplish this there is some special logic surrounding what to show next.
     */
    let next = articlesThatArentSecret.slice(index + 1, index + 3);
    // If it's the last item in the list, there will be no articles. So grab the first 2
    if (next.length === 0) next = articlesThatArentSecret.slice(0, 2);
    // If there's 1 item in the list, grab the first article
    if (next.length === 1 && articlesThatArentSecret.length !== 2)
      next = [...next, articlesThatArentSecret[0]];
    if (articlesThatArentSecret.length === 1) next = [];

    createPage({
      path: article.slug,
      component: templates.article,
      context: {
        article,
        authors: authorsThatWroteTheArticle,
        tags: tagsThatWroteTheArticle,
        basePath,
        permalink: `${data.site.siteMetadata.siteUrl}${article.slug}/`,
        slug: article.slug,
        id: article.id,
        title: article.title,
        canonicalUrl: article.canonical_url,
        mailchimp,
        next,
      },
    });
  });

  /**
   * By default the author's page is not enabled. This can be enabled through the theme options.
   * If enabled, each author will get their own page and a list of the articles they have written.
   */
  if (authorsPage) {
    log('Creating', 'authors page');

    authors.forEach(author => {
      const articlesTheAuthorHasWritten = articlesThatArentSecret.filter(
        article =>
          article.author.toLowerCase().includes(author.name.toLowerCase()),
      );
      const path = slugify(author.slug, authorsPath);

      createPaginatedPages({
        edges: articlesTheAuthorHasWritten,
        pathPrefix: author.slug,
        createPage,
        pageLength,
        pageTemplate: templates.author,
        buildPath: buildPaginatedPath,
        context: {
          author,
          originalPath: path,
          skip: pageLength,
          limit: pageLength,
        },
      });
    });
  }

  /**
   * By default the tag's page is not enabled. This can be enabled through the theme options.
   * If enabled, each tag will get their own page and a list of the articles they have written.
   */
  if (tagsPage) {
    log('Creating', 'tags page');

    tags.forEach(tag => {
      const articlesTheTagHasWritten = articlesThatArentSecret.filter(
        article =>
          article.tag.toLowerCase().includes(tag.name.toLowerCase()),
      );
      const path = slugify(tag.slug, tagsPath);

      createPaginatedPages({
        edges: articlesTheTagHasWritten,
        pathPrefix: tag.slug,
        createPage,
        pageLength,
        pageTemplate: templates.tag,
        buildPath: buildPaginatedPath,
        context: {
          tag,
          originalPath: path,
          skip: pageLength,
          limit: pageLength,
        },
      });
    });
  }
};
